src/utils/hooks/useCharacterCount/useCharacterCount.ts

```diff
@@ -0,0 +1,77 @@
+import { useReducer, useCallback } from 'react';
+
+import {
+  reducer,
+  CharacterCountState,
+  UserAction,
+} from './useCharacterCount.state';
+
+export interface UseCharacterCountInit {
+  value?: string;
+  /** Max number of characters allowed */
+  maxCharacters?: number;
+  /** Shows warning color when threshold is met */
+  warningThreshold?: number;
+}
+
+export interface UseCharacterCount {
+  state: CharacterCountState;
+  dispatch: React.Dispatch<UserAction>;
+  handleChange: (value: string) => void;
+  clean: () => void;
+}
+
+/**
+ * Based on withCharacterCount, this provides handlers for a character count message to create a controlled component.
+ * Helpful to use in conjuction with custom inputs that need more control of input messages
+ */
+
+export function useCharacterCount({
+  maxCharacters = 10,
+  warningThreshold = 5,
+}: UseCharacterCountInit): UseCharacterCount {
+  const [state, dispatch] = useReducer(reducer, {
+    remainingCharacters: maxCharacters,
+  });
+
+  function handleError() {
+    dispatch({
+      type: 'SET_ERROR',
+    });
+  }
+
+  function handleWarn() {
+    dispatch({
+      type: 'SET_WARNING',
+    });
+  }
+
+  const clean = useCallback(() => {
```

Comment:
Curious why is only this action wrapped in useCalkback?

Reply:
Ah, I added this because of an eslint error: 

```
The 'clean' function makes the dependencies of useCallback Hook (at line 68) change on every render. To fix this, wrap the definition of 'clean' in its own useCallback() Hook.
```